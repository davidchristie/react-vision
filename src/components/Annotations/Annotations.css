.Annotations {
  padding: 10px;
}
