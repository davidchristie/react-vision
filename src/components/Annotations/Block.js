import PropTypes from 'prop-types'
import React, { Component } from 'react'

import './Block.css'
import Paragraph from './Paragraph'

export default class Block extends Component {
  static propTypes = {
    paragraphs: PropTypes.arrayOf(
      PropTypes.object
    ).isRequired
  }

  render () {
    const { paragraphs } = this.props
    return (
      <div className='Block'>
        {
          paragraphs.map((paragraph, index) => {
            return <Paragraph key={index} {...paragraph} />
          })
        }
      </div>
    )
  }
}
