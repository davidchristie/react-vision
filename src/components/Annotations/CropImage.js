import PropTypes from 'prop-types'
import React, { Component } from 'react'

const { Image } = window

export default class CropImage extends Component {
  static propTypes = {
    image: PropTypes.string.isRequired,
    vertices: PropTypes.arrayOf(
      PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number
      })
    ).isRequired
  }

  componentDidMount () {
    this.drawImage()
  }

  componentWillReceiveProps () {
    this.drawImage()
  }

  drawImage () {
    const { image, vertices } = this.props
    const { canvas } = this.refs
    var img = new Image()
    img.addEventListener('load', () => {
      canvas.height = img.height
      canvas.width = img.width
      const ctx = canvas.getContext('2d')
      ctx.drawImage(img, 0, 0, canvas.width, canvas.height)
      const x = vertices[0].x || 0
      const y = vertices[0].y || 0
      const width = vertices[2].x
      const height = vertices[2].y
      ctx.strokeStyle = 'red'
      ctx.lineWidth = 5
      ctx.rect(x, y, width, height)
      ctx.stroke()
    })
    img.setAttribute('src', image)
  }

  render () {
    return <canvas ref='canvas' />
  }
}
