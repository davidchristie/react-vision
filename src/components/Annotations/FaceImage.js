import PropTypes from 'prop-types'
import React, { Component } from 'react'

import ImageOverlay from './ImageOverlay'

export default class FaceImage extends Component {
  static propTypes = {
    image: PropTypes.string.isRequired,
    face: PropTypes.shape({
      boundingPoly: PropTypes.shape({
        vertices: PropTypes.arrayOf(
          PropTypes.shape({
            x: PropTypes.number,
            y: PropTypes.number
          })
        ).isRequired
      }).isRequired
    }).isRequired
  }

  constructor (props) {
    super(props)
    this.drawOverlay = this.drawOverlay.bind(this)
  }

  drawOverlay (ctx) {
    const {
      face: {
        boundingPoly: {
          vertices
        },
        landmarks
      }
    } = this.props
    const x = vertices[0].x || 0
    const y = vertices[0].y || 0
    const width = vertices[2].x - x
    const height = vertices[2].y - y
    ctx.fillStyle = 'red'
    ctx.strokeStyle = 'red'
    ctx.lineWidth = 5
    ctx.rect(x, y, width, height)
    landmarks.forEach(landmark => {
      const position = landmark.position
      const { x, y } = position
      ctx.fillRect(x - 3, y - 3, 6, 6)
    })
    ctx.stroke()
  }

  render () {
    const { image } = this.props
    return (
      <ImageOverlay
        image={image}
        overlay={this.drawOverlay}
      />
    )
  }
}
