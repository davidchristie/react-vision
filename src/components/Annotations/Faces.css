.Faces > div {
  border: thin solid;
  margin: 10px;
  padding: 10px;
}
