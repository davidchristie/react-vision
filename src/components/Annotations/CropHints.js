import PropTypes from 'prop-types'
import React, { Component } from 'react'

import CropImage from './CropImage'
import './CropHints.css'

export default class CropHints extends Component {
  static propTypes = {
    cropHints: PropTypes.arrayOf(
      PropTypes.shape({
        boundingPoly: PropTypes.shape({
          vertices: PropTypes.arrayOf(
            PropTypes.shape({
              x: PropTypes.number,
              y: PropTypes.number
            })
          ).isRequired
        }).isRequired,
        confidence: PropTypes.number.isRequired,
        importanceFraction: PropTypes.number.isRequired
      }).isRequired
    ).isRequired,
    image: PropTypes.string.isRequired
  }

  render () {
    const { cropHints, image } = this.props
    return (
      <div className='CropHints'>
        <h3>Crop Hints</h3>
        {
          cropHints.map((cropHint, index) => {
            const {
              boundingPoly: {
                vertices
              },
              confidence,
              importanceFraction
            } = cropHint
            return (
              <div key={index}>
                <CropImage image={image} vertices={vertices} />
                <p>
                  <strong>Bounding polygon: </strong>
                  {
                    vertices
                      .map((vertex, index) => {
                        const x = vertex.x || 0
                        const y = vertex.y || 0
                        return `(${x}, ${y})`
                      })
                      .join(', ')
                  }
                </p>
                <p>
                  <strong>Confidence: </strong>
                  {confidence}
                </p>
                <p>
                  <strong>Importance fraction: </strong>
                  {importanceFraction}
                </p>
              </div>
            )
          })
        }
      </div>
    )
  }
}
