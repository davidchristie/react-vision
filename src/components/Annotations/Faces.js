import PropTypes from 'prop-types'
import React, { Component } from 'react'

import './Faces.css'
import FaceImage from './FaceImage'

export default class Faces extends Component {
  static propTypes = {
    faces: PropTypes.arrayOf(
      PropTypes.shape({
        angerLikelihood: PropTypes.string.isRequired,
        blurredLikelihood: PropTypes.string.isRequired,
        detectionConfidence: PropTypes.number.isRequired,
        headwearLikelihood: PropTypes.string.isRequired,
        joyLikelihood: PropTypes.string.isRequired,
        landmarks: PropTypes.arrayOf(
          PropTypes.shape({
            position: PropTypes.shape({
              x: PropTypes.number.isRequired,
              y: PropTypes.number.isRequired
            })
          })
        ).isRequired,
        sorrowLikelihood: PropTypes.string.isRequired,
        surpriseLikelihood: PropTypes.string.isRequired,
        underExposedLikelihood: PropTypes.string.isRequired
      })
    ).isRequired,
    image: PropTypes.string.isRequired
  }

  render () {
    const { faces, image } = this.props
    return (
      <div className='Faces'>
        <h3>Faces</h3>
        {
          faces.map((face, index) => {
            return (
              <div key={index}>
                <FaceImage
                  face={face}
                  image={image}
                />
                <p>
                  <strong>Anger: </strong>
                  {face.angerLikelihood}
                </p>
                <p>
                  <strong>Joy: </strong>
                  {face.joyLikelihood}
                </p>
                <p>
                  <strong>Sorrow: </strong>
                  {face.sorrowLikelihood}
                </p>
                <p>
                  <strong>Surprise: </strong>
                  {face.surpriseLikelihood}
                </p>
                <p>
                  <strong>Headwear: </strong>
                  {face.headwearLikelihood}
                </p>
                <p>
                  <strong>Blurred: </strong>
                  {face.blurredLikelihood}
                </p>
                <p>
                  <strong>Under Exposed: </strong>
                  {face.underExposedLikelihood}
                </p>
                <p>
                  <strong>Detection Confidence: </strong>
                  {face.detectionConfidence}
                </p>
              </div>
            )
          })
        }
      </div>
    )
  }
}
