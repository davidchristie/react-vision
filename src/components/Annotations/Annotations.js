import PropTypes from 'prop-types'
import React, { Component } from 'react'

import './Annotations.css'
import CropHints from './CropHints'
import Faces from './Faces'
import FullText from './FullText'
import ImageProperties from './ImageProperties'
import Labels from './Labels'
import SafeSearch from './SafeSearch'
import Tabs from './Tabs'
import Web from './Web'

export default class Annotations extends Component {
  static propTypes = {
    data: PropTypes.object,
    image: PropTypes.string.isRequired
  }

  render () {
    const { data, image } = this.props
    if (!data) {
      return (
        <div className='Annotations'>
          <h2>Loading...</h2>
        </div>
      )
    }
    if (!data.responses) {
      return (
        <div className='Annotations'>
          <h2>Network error</h2>
        </div>
      )
    }
    const response = data.responses[0]
    console.log(response)
    const {
      cropHintsAnnotation,
      faceAnnotations,
      fullTextAnnotation,
      labelAnnotations,
      imagePropertiesAnnotation,
      safeSearchAnnotation,
      webDetection
    } = response
    const tabs = []
    if (cropHintsAnnotation) {
      tabs.push({
        content: <CropHints image={image} {...cropHintsAnnotation} />,
        name: 'Crop Hints'
      })
    }
    if (faceAnnotations) {
      tabs.push({
        content: <Faces faces={faceAnnotations} image={image} />,
        name: 'Faces'
      })
    }
    if (labelAnnotations) {
      tabs.push({
        content: <Labels labels={labelAnnotations} />,
        name: 'Labels'
      })
    }
    if (imagePropertiesAnnotation) {
      tabs.push({
        content: <ImageProperties imagePropertiesAnnotation={imagePropertiesAnnotation} />,
        name: 'Properties'
      })
    }
    if (safeSearchAnnotation) {
      tabs.push({
        content: <SafeSearch safeSearch={safeSearchAnnotation} />,
        name: 'Safe Search'
      })
    }
    if (fullTextAnnotation) {
      tabs.push({
        content: <FullText fullText={fullTextAnnotation} />,
        name: 'Text'
      })
    }
    if (webDetection) {
      tabs.push({
        content: <Web web={webDetection} />,
        name: 'Web'
      })
    }
    return (
      <div className='Annotations'>
        <Tabs tabs={tabs} />
      </div>
    )
  }
}
