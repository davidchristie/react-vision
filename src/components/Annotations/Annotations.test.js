import { shallow } from 'enzyme'
import React from 'react'

import Annotations from './Annotations'

it('renders correctly', () => {
  const props = {
    image: 'imageDataUrl'
  }
  expect(shallow(<Annotations {...props} />)).toMatchSnapshot()
})
