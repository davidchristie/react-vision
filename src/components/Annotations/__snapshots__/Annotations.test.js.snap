// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Annotations"
>
  <h2>
    Loading...
  </h2>
</div>
`;
